File: ShapeByte/framework/device/device_size/tests/DeviceSizeCategoryTest.swift
//
//  DeviceSizeCategoryTest.swift
//  ShapeByteTests
//
//  Created by Lang, Stefan [RTL Tech] on 13.08.24.
//

import XCTest

final class DeviceSizeCategoryTest: XCTestCase {

    override func setUpWithError() throws {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDownWithError() throws {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
    }

    func testSize() throws {
        var size: CGSize = CGSize(width: 320, height: 568)
        var sut: DeviceSizeCategory = DeviceSizeCategory.xSmall(size: size)
        XCTAssertEqual(size, sut.size)

        size = CGSize(width: 375, height: 667)
        sut = DeviceSizeCategory.small(size: size)
        XCTAssertEqual(size, sut.size)

        size = CGSize(width: 390, height: 844)
        sut = DeviceSizeCategory.medium(size: size)
        XCTAssertEqual(size, sut.size)

        size = CGSize(width: 430, height: 932)
        sut = DeviceSizeCategory.medium(size: size)
        XCTAssertEqual(size, sut.size)
    }

    func testIsXSmall() throws {
        let size: CGSize = CGSize(width: 320, height: 568)
        let sut: DeviceSizeCategory = DeviceSizeCategory.xSmall(size: size)

        XCTAssertTrue(sut.isXSmall)
        XCTAssertFalse(sut.isSmall)
        XCTAssertFalse(sut.isMedium)
        XCTAssertFalse(sut.isLarge)
    }

    func testIsSmall() throws {
        let size: CGSize = CGSize(width: 320, height: 667)
        let sut: DeviceSizeCategory = DeviceSizeCategory.small(size: size)

        XCTAssertFalse(sut.isXSmall)
        XCTAssertTrue(sut.isSmall)
        XCTAssertFalse(sut.isMedium)
        XCTAssertFalse(sut.isLarge)
    }

    func testIsMedium() throws {
        let size: CGSize = CGSize(width: 390, height: 844)
        let sut: DeviceSizeCategory = DeviceSizeCategory.medium(size: size)

        XCTAssertFalse(sut.isXSmall)
        XCTAssertFalse(sut.isSmall)
        XCTAssertTrue(sut.isMedium)
        XCTAssertFalse(sut.isLarge)
    }

    func testIsLarge() throws {
        let size: CGSize = CGSize(width: 430, height: 932)
        let sut: DeviceSizeCategory = DeviceSizeCategory.large(size: size)

        XCTAssertFalse(sut.isXSmall)
        XCTAssertFalse(sut.isSmall)
        XCTAssertFalse(sut.isMedium)
        XCTAssertTrue(sut.isLarge)
    }
}

File: shared/foundation/core/api/src/commonTest/kotlin/de/stefan/lang/foundationCore/api/devicesize/DeviceSizeCategoryTest.kt
package de.stefan.lang.foundationCore.api.devicesize

import kotlin.test.Test
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class DeviceSizeCategoryTest {
    private val xSmall = DeviceSizeCategory.XSmall(Size(100f, 200f))
    private val small = DeviceSizeCategory.Small(Size(300f, 500f))
    private val medium = DeviceSizeCategory.Medium(Size(500f, 750f))
    private val large = DeviceSizeCategory.Large(Size(800f, 1280f))

    @Test
    fun `isXSmall returns true for XSmall`() {
        assertTrue(xSmall.isXSmall)
        assertFalse(small.isXSmall)
        assertFalse(medium.isXSmall)
        assertFalse(large.isXSmall)
    }

    @Test
    fun `isSmall returns true for Small`() {
        assertFalse(xSmall.isSmall)
        assertTrue(small.isSmall)
        assertFalse(medium.isSmall)
        assertFalse(large.isSmall)
    }

    @Test
    fun `isMedium returns true for Medium`() {
        assertFalse(xSmall.isMedium)
        assertFalse(small.isMedium)
        assertTrue(medium.isMedium)
        assertFalse(large.isMedium)
    }

    @Test
    fun `isLarge returns true for Large`() {
        assertFalse(xSmall.isLarge)
        assertFalse(small.isLarge)
        assertFalse(medium.isLarge)
        assertTrue(large.isLarge)
    }
}
